staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -330,7 +330,17 @@ func patchResource(
 				return nil, err
 			}
 
-			hasConflicts, err := mergepatch.HasConflicts(originalPatchMap, currentPatchMap)
+			var hasConflicts bool
+			switch patchType {
+			case types.StrategicMergePatchType:
+				patchMetaFromStruct, e := strategicpatch.NewPatchMetaFromStruct(versionedObj)
+				if e != nil {
+					return nil, e
+				}
+				hasConflicts, err = strategicpatch.MergingMapsHaveConflicts(originalPatchMap, currentPatchMap, patchMetaFromStruct)
+			case types.JSONPatchType, types.MergePatchType:
```

Comment:
this is not right... getOriginalPatchMap() always returns a strategic patch, regardless of the original patch type.

Reply:
Are you saying that we should use `strategicpatch.MergingMapsHaveConflicts(...)` for all patch types?
It was using `mergepatch.HasConflicts(...)` before.

Reply:
if that is the correct way to compare strategic merge patches, then yes.

separately, it's worth revisiting the assumption that a strategic merge patch can be computed for all types, and to possibly begin using merge patches for conflict detection if a strategic merge patch cannot be computed.

in any case, the submitted `patchType` is currently unrelated to the patch compared here for conflicts.

Reply:
>if that is the correct way to compare strategic merge patches, then yes.

You are right that `getOriginalPatchMap()` always returns a SMP. Then I think we should use `strategicpatch.MergingMapsHaveConflicts(...)` for all patch types.
IMO the reason is that `mergepatch.HasConflicts(...)` is too restrict for 2 SMPs, it requires:
- the list to have same length.
- items at the same index have no conflict.
It don't use MergeKey to identify the entries.

IMO it makes more sense to use `strategicpatch.MergingMapsHaveConflicts(...)` .

> separately, it's worth revisiting the assumption that a strategic merge patch can be computed for all types, and to possibly begin using merge patches for conflict detection if a strategic merge patch cannot be computed.

SMP won't work for unknown types e.g. CRDs. IOW current `getOriginalPatchMap()` may not always succeed.

Reply:
I want to understand how conflict could happen in the APIserver. IIUC PATCH goes to the same path as PUT to update the object. The conflict can only caused by optimistic locking (Resource Version). Is this understanding correct?